Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,272 @@
+            {
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
+            }
+        }
+
+        _writeProperties = writeProperties.ToArray();
+        _readProperties = readProperties.ToArray();
+
+        _readPropertiesByName = readProperties
+            .SelectMany(p => p.ReadNames.Select((n, i) => (IsPrimary: i == 0, n, DTOProperty: p)))
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions GetPropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
+    {
+        if (converter == null)
+        {
+            return options;
+        }
+
+        var cloned = new JsonSerializerOptions(options);
+        cloned.Converters.Insert(0, converter);
+
+        return cloned;
+    }
+
+    /// <inheritdoc />
+    public override TInterface Read
+    (
+        ref Utf8JsonReader reader,
+        Type typeToConvert,
+        JsonSerializerOptions options
+    )
+    {
+        if (reader.TokenType != JsonTokenType.StartObject ||
+            !reader.Read())
+        {
+            DataObjectConverterShared.ThrowJsonException();
+        }
+
+        var readProperties = _readProperties;
+
+        // CMBK: Considering to cache this array and reuse it.
+        // F.e. in a thread-static field. This method isn't async anyways.
+
+        // Avoid creating a dictionary and fill the arguments array directly
+        var constructorArguments = new object?[readProperties.Length];
+
+        // We use "Missing" as a sentinel to indicate that a value hasn't
+        // been seen yet without needing to allocate additional memory.
+        constructorArguments.AsSpan().Fill(DataObjectConverterShared.Missing);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            var propertyName = reader.GetString();
+            if (!reader.Read())
+            {
+                DataObjectConverterShared.ThrowJsonException();
+            }
+
+            Debug.Assert(propertyName != null, "Expected property names to never be null.");
+            var (isPrimaryChoice, dtoProperty) = GetReadPropertyInfo(propertyName);
+
+            if (dtoProperty is null)
+            {
+                if (!_allowExtraProperties)
+                {
+                    DataObjectConverterShared.ThrowJsonException();
+                }
+
+                // No matching property - we'll skip it
+                if (!reader.TrySkip())
+                {
+                    DataObjectConverterShared.ThrowJsonException("Couldn't skip elements.");
+                }
+
+                if (!reader.Read())
+                {
+                    DataObjectConverterShared.ThrowJsonException
+                    (
+                        $"No matching DTO property for JSON property \"{propertyName}\" could be found."
```

Comment:
Strange error message? Is this condition not hit when a) we allow extra properties, b) we could skip the property, and c) we reached the end of the JSON?

Reorganize surrounding conditions or fix message to clarify.

Reply:
This error message (and the entire surrounding block when the property wasn't found for that matter) was copied verbatim. Sounds like it should be "Unexpectedly reached end of JSON." I'll modify it.